### ISSUE TITLE
ASPM support?

### ISSUE THREAD

**Issue by j4cbo** [google-coral/edgetpu#834]
### Description

I have an M.2 A+E key Coral module installed in a Lenovo M920q PC running TrueNAS, with the latest BIOS and all tuning suggestions from `powertop` applied.

Without the TPU installed, the PC can spend most of its time in package C-state PC8, but with the TPU present and fully idle, the PC doesn't go below C-state PC3. Simply having the TPU installed adds about 3 watts to the PC's idle power consumption. This is significantly above the datasheet 375-400mW idle power for the TPU itself, and I suspect can mostly be attributed to the PC being unable to enter lower C-states. `lspci -vvv` reports that the TPU supports ASPM L0s/L1 but that it's disabled, and the same appears with `pcie_aspm=force` on the kernel command line.

I've seen a few references to manually disabling ASPM in other bug reports, but I'm not sure if that's as a workaround for ASPM bugs in host systems or due to issues with the Coral device itself. So... is this expected to work and the issue is some kernel or BIOS misconfiguration on my side, or does the Coral just not support ASPM?

<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Operating System

Linux

### Coral Device

M.2 Accelerator A+E

### Programming Language

Other

### Relevant Log Output

_No response_</details>

**Comment by vrecky** [comment]
Hi, is that a tiny pc ?

**Comment by antipesto93** [comment]
Same issue on a similar tiny pc, ASPM disabled causes CPU to stay in low cstate, double the idle power consumption of the host

**Comment by VMmatty** [comment]
Thought I'd add in I've got the same issue.  M.2 Coral being used in Frigate in a container running on Unraid.  With the TPU installed, even if Frigate isn't running, still forces the system to stay in C3.  As soon as I remove the Coral I can go as low as C8 and the power consumption drops over 10W.

Would love to figure out a way to get lower than C3 and still use the Coral.

**Comment by chiluk** [comment]
Have any of you discovered a workaround or fix for this?

**Comment by j4cbo** [comment]
I never did. I wound up switching my Frigate setup from Coral to OpenVINO, but that's not much of a workaround 🙁

**Comment by sbidy** [comment]
I tried to force the TPU to ASPM but this seems to break the whole system.

Do you get it done?

**Comment by Washee** [comment]
Is there any solution to this problem?
Are there TPUs out there with better driver support?

**Comment by thefl0yd** [comment]
> Is there any solution to this problem? Are there TPUs out there with better driver support?

No, this project is effectively dead. Google has abandoned it. 

The only support here is community support and the community can’t fix the hardware. There’s a handful of hardware related bugs that have been reported and pinged on for 2+ years and google has shown zero interest in replying. 

Unfortunately you need to move on to something else or accept the TPU more or less as is.

**Comment by VMmatty** [comment]
thefl0yd is correct, throw the Coral in the pile of abandoned Google projects.  I was using it for object detection in Frigate but abandoned it in favor of the iGPU and it works just as well with only a few milliseconds higher inference time.
